### PR TITLE
Improve Validation on Problem Endpoints

### LINKED ISSUE #17: Validation in problem create/update route
Check if referenced problemSet exists before creating the problem then send back an appropriate error message.

### PR DESCRIPTION
**Pull request by idodin**
Resolves #17 #20

### CODE REVIEW

File: src/validators/problemCustom.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+export async function validateProblemSetIds(problemSetIds: Array<string>) {
+    const promiseArray = problemSetIds.map(problemSetId => problemSetDBInteractions.find(problemSetId));
+    const problemSetArray = await Promise.all(promiseArray);
+    return !problemSetArray.some(problemSet => problemSet === null);
+}
+
+export function validateProblemMetadata(problemMetadata) {
+    return !(!("platformProblemId" in problemMetadata)
+        || !("difficulty" in problemMetadata)
+        || !(problemMetadata.difficulty in Difficulty));
```
Comment (ValerianClerc) [inline src/validators/problemCustom.ts:13]: We can use DeMorgan's Law to simplify this boolean logic :P

File: src/controllers/problem.ts [hunk 2]
```diff
@@ -101,7 +101,7 @@ const problemController = {
                     };
 
 
-                    if (req.body.source && req.body.problemMetadata && req.body.problemMetadata.platformProblemId) {
+                    if (req.body.source && req.body.problemMetadata) {
```
Comment (MohamedBeydoun) [inline src/controllers/problem.ts:104]: can be removed since they are required in the validator

File: src/validators/problem.ts [hunk 3]
```diff
@@ -28,7 +30,10 @@ export function problemValidator(method: string): ValidationChain[] {
                 body("source", "Invalid or missing 'source'").isString().exists(),
                 body("sourceLink", "Invalid or missing 'sourceLink'").isString().exists(),
                 body("problemSetIds", "Invalid or missing 'problemSetId'").isArray().exists(),
-                body("problemMetadata", "Invalid or missing 'problemMetadata' (optional)").optional()
+                body("problemSetIds", "Invalid or nonexistent 'problemSetId'").custom(validateProblemSetIds),
+                body("problemMetadata", "Invalid or missing fields in 'problemMetadata' (optional)")
+                    .optional()
```
Comment (ValerianClerc) [inline src/validators/problem.ts:35]: should our PUT request have optional parameters? I think the right way to do PUT is for the client to send the full object body, same as in a POST request

File: src/controllers/problem.ts (hunk 2 above)
Comment (ValerianClerc) [inline src/controllers/problem.ts:104]: See my comment about PUT requests (we won't need this check if we require these fields in validation)

File: src/validators/problem.ts [hunk 4]
```diff
@@ -28,7 +30,10 @@ export function problemValidator(method: string): ValidationChain[] {
...
+                    .custom(validateProblemMetadata)
```
Comment (MohamedBeydoun) [inline src/validators/problem.ts:36]: PUT should be the same as POST (rfc conventions)

File: src/controllers/problem.ts (hunk 2 above)
Comment (ValerianClerc) [reply src/controllers/problem.ts:104]: only source is required in validator. See my comments

File: src/validators/problemCustom.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+import { Difficulty, PlatformName } from "../interfaces/IProblem";
+
+export async function validateProblemSetIds(problemSetIds: Array<string>) {
+    const promiseArray = problemSetIds.map(problemSetId => problemSetDBInteractions.find(problemSetId));
+    const problemSetArray = await Promise.all(promiseArray);
+    return !problemSetArray.some(problemSet => problemSet === null);
+}
+
+export function validateProblemMetadata(problemMetadata) {
+    console.log(problemMetadata.difficulty)
```
Comment (MohamedBeydoun) [inline src/validators/problemCustom.ts:11]: remove console.log()

File: src/validators/problem.ts [hunk 6]
```diff
@@ -14,21 +14,23 @@ export function problemValidator(method: string): ValidationChain[] {
...
             return [
                 param("problemId", "Invalid or missing ':problemId'").exists().isMongoId(),
                 body("title", "Invalid or missing 'title'").isString().exists(),
-                body("source", "Invalid or missing 'source'").isString().exists(),
+                body("source", "Invalid or missing 'source'").isString().exists().custom(validateProblemPlatform),
                 body("sourceLink", "Invalid or missing 'sourceLink'").isString().exists(),
                 body("problemSetIds", "Invalid or missing 'problemSetId'").isArray().exists(),
-                body("problemMetadata", "Invalid or missing 'problemMetadata' (optional)").optional()
+                body("problemSetIds", "Invalid or nonexistent 'problemSetId'").custom(validateProblemSetIds),
+                body("problemMetadata", "Invalid or missing fields in 'problemMetadata'").custom(validateProblemMetadata)
```
Comment (ValerianClerc) [inline src/validators/problem.ts:33]: should we add `.exists()` checks for these custom ones too?

File: src/controllers/problem.ts [hunk 7]
```diff
@@ -95,20 +99,21 @@ const problemController = {
                 if (!problem)
                     res.status(statusCodes.NOT_FOUND).send({ status: statusCodes.NOT_FOUND, message: "Problem not found" });
                 else {
+                    const { problemMetadata } = req.body.problemMetadata;
```
Comment (ValerianClerc) [inline src/controllers/problem.ts:102]: This destructuring should be `const {problemMetadata} = req.body`

File: src/controllers/problem.ts [hunk 8]
```diff
@@ -56,12 +56,16 @@ const problemController = {
             res.status(statusCodes.MISSING_PARAMS).json(errors.formatWith(errorMessage).array()[0]);
         } else {
             try {
-                let platformProblemId = req.body.problemMetadata.platformProblemId
-                const platform = req.body.source.toUpperCase()
+                const { problemMetadata } = req.body.problemMetadata;
```
Comment (ValerianClerc) [inline src/controllers/problem.ts:59]: See destructuring object comment

File: src/validators/problem.ts [hunk 9]
```diff
@@ -14,21 +14,23 @@ export function problemValidator(method: string): ValidationChain[] {
...
             return [
                 body("title", "Invalid or missing 'title'").isString().exists(),
-                body("source", "Invalid or missing 'source'").isString().exists(),
+                body("source", "Invalid or missing 'source'").isString().exists().custom(validateProblemPlatform),
                 body("sourceLink", "Invalid or missing 'sourceLink'").isString().exists(),
                 body("problemSetIds", "Invalid or missing 'problemSetId'").isArray().exists(),
                 body("problemSetIds", "Invalid or nonexistent 'problemSetId'").custom(validateProblemSetIds),
-                body("problemMetadata", "Invalid or missing 'problemMetadata' (optional)").optional()
+                body("problemMetadata", "Missing 'problemMetadata'").exists(),
+                body("problemMetadata", "Invalid or missing fields in 'problemMetadata'").custom(validateProblemMetadata)
```
Comment (ValerianClerc) [inline src/validators/problem.ts:22]: exists() here?

File: src/validators/problem.ts (hunk 6 above)
Comment (ValerianClerc) [reply src/validators/problem.ts:33]: For `problemSetIds`, why have two different lines of validations returning the same message? The custom one should return a more specific message if we split them like that.

File: src/validators/problemCustom.ts [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+
+export function validateProblemMetadata(problemMetadata) {
+    return ("platformProblemId" in problemMetadata)
+        && ("difficulty" in problemMetadata)
+        && (problemMetadata.difficulty.toLowerCase() in Difficulty);
+}
+
+export function validateProblemPlatform(platform) {
+    return platform.toUpperCase() in PlatformName;
+}
```
Comment (MohamedBeydoun) [inline src/validators/problemCustom.ts:18]: could we change the format of the function declarations to match what we have in other files?
e.g.
export const validateProblemPlatform = (platform) => {
    ...
}

File: src/validators/problem.ts (hunk 6 above)
Comment (idodin) [reply src/validators/problem.ts]: They are different messages 'Invalid or Missing' vs 'Invalid or nonexistent'

File: src/controllers/problem.ts [hunk 11]
```diff
@@ -95,20 +99,21 @@ const problemController = {
...
 
-                    if (req.body.source && req.body.problemMetadata && req.body.problemMetadata.platformProblemId) {
-                        let platformProblemId = req.body.problemMetadata.platformProblemId
-                        const platform = req.body.source.toUpperCase()
-                        if (platform == "CODEFORCES") {
-                            platformProblemId = platformProblemId.toUpperCase();
-                        }
-                        updatedProblemBody.problemId = calculateProblemHash(platform, platformProblemId);
+                    let platformProblemId = problem.problemMetadata.platformProblemId
+                    const platform = problem.source;
```
Comment (MohamedBeydoun) [inline src/controllers/problem.ts:112]: `const problem` is already defined. Change with:
```javascript
let updatedProblemBody: IProblem = {
    ...req.body,
    source: req.body.source.toUpperCase()
}
updatedProblemBody.problemMetadata.difficulty = updatedProblemBody.problemMetadata.difficulty.toLowerCase();

let platformProblemId = updatedProblemBody.problemMetadata.platformProblemId;
const platform = updatedProblemBody.source;
```

**Pull request merged**